flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,672 @@
+        WithPreCommitTopology<RowData, FilesCommittable>,
+        WithPostCommitTopology<RowData, FilesCommittable> {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSink.class);
+
+  private final TableLoader tableLoader;
+  private final Table table;
+  private final boolean overwrite;
+  private final boolean upsertMode;
+  private final DistributionMode distributionMode;
+  private final List<String> equalityFieldColumns;
+  private final List<Integer> equalityFieldIds;
+  private final FileFormat dataFileFormat;
+  private final int workerPoolSize;
+  private final long targetDataFileSize;
+  private final String uidPrefix;
+  private final Map<String, String> snapshotProperties;
+  private final RowType flinkRowType;
+  private Committer<FilesCommittable> committer;
+
+  public IcebergSink(
+      TableLoader tableLoader,
+      @Nullable Table newTable,
+      TableSchema tableSchema,
+      List<String> equalityFieldColumns,
+      @Nullable String uidPrefix,
+      ReadableConfig readableConfig,
+      Map<String, String> writeOptions,
+      Map<String, String> snapshotProperties) {
+    this.tableLoader = tableLoader;
+    this.equalityFieldColumns = equalityFieldColumns;
+    this.uidPrefix = uidPrefix == null ? "" : uidPrefix;
+    this.snapshotProperties = snapshotProperties;
+
+    Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
+
+    if (newTable == null) {
+      tableLoader.open();
+      try (TableLoader loader = tableLoader) {
+        this.table = loader.loadTable();
+      } catch (IOException e) {
+        throw new UncheckedIOException(
+            "Failed to load iceberg table from table loader: " + tableLoader, e);
+      }
+    } else {
+      this.table = newTable;
+    }
+
+    FlinkWriteConf flinkWriteConf = new FlinkWriteConf(table, writeOptions, readableConfig);
+
+    this.equalityFieldIds = checkAndGetEqualityFieldIds(table, equalityFieldColumns);
+
+    this.flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+
+    // Validate the equality fields and partition fields if we enable the upsert mode.
+    if (flinkWriteConf.upsertMode()) {
+      Preconditions.checkState(
+          !flinkWriteConf.overwriteMode(),
+          "OVERWRITE mode shouldn't be enable when configuring to use UPSERT data stream.");
+      Preconditions.checkState(
+          !equalityFieldIds.isEmpty(),
+          "Equality field columns shouldn't be empty when configuring to use UPSERT data stream.");
+      if (!table.spec().isUnpartitioned()) {
+        for (PartitionField partitionField : table.spec().fields()) {
+          Preconditions.checkState(
+              equalityFieldIds.contains(partitionField.sourceId()),
+              "In UPSERT mode, partition field '%s' should be included in equality fields: '%s'",
+              partitionField,
+              equalityFieldColumns);
+        }
+      }
+    }
+
+    this.upsertMode = flinkWriteConf.upsertMode();
```

Comment:
We probably should make `FlinkWriteConf` serializable to avoid copying the config values.

Reply:
I tried it once, bug `FlinkConfParser` `ReadableConfig` ... blocked me, I didn't go that way. Or we could migrate it inside the `Builder`, as SparkWriberBuilder does？
https://github.com/apache/iceberg/blob/263441752393834c384a04d861cda1b8cb136a63/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkWriteBuilder.java#L76-L86

Reply:
we can extract the config inside the `FlinkWriteConf`, which can be serializable. we don't have to make `FlinkConfParser` serializable, as `ReadableConfig` is not serializable.

Reply:
FlinkWriteConf uses `FlinkConfParser` and `ReadableConfig` internally, serializing `FlinkWriteConf` is not enough.

Reply:
`FlinkWriteConf` doesn't have to keep the `FlinkConfParser` as class members. Instead, `FlinkWriteConf` can use a local `FlinkConfParser` variable to extract the individual config values (similar to what you did here) and only save the individual config values (not `FlinkConfParser`) as class members

Reply:
These  config values are retrieved from `Table\Config\...` each time they are used. Do I assign all of them in advance at creation time?? I don't recommend doing that.
For example, if we did that and the user changed the value of the table or config after initializing FlinkWriteConfig, then we would get the wrong value.(Of course, in our case, it doesn't happen, it's read-only.)

Reply:
> Do I assign all of them in advance at creation time

Reply:
> your current implementation also only retrieve the config values once during job startup/initialization. there is no difference

Yes, I've assigned all the values here, but at this stage, we've entered the job parsing phase, and we can assume that we're no longer receiving changes.

However, we do not provide the same guarantees when initializing FlinkWriteConf. We only get the necessary values when needed, and it ensure that the value is ultimately correct.

It's my considered opinion that we'd better not serialize them. In fact, when we created them, we mentioned in doc that these classes would not be serialized.

https://github.com/apache/iceberg/blob/af4b39555374e1d1ab966278c5bc94ea41ba91af/flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkWriteConf.java#L43-L49